org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/quickfix/WorkbenchMarkerResolutionAdapter.xtend

```diff
@@ -0,0 +1,130 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.ui.editor.quickfix
+
+import com.google.inject.Inject
+import com.google.inject.Provider
+import java.lang.reflect.InvocationTargetException
+import org.apache.log4j.Logger
+import org.eclipse.core.resources.IMarker
+import org.eclipse.core.runtime.CoreException
+import org.eclipse.core.runtime.IProgressMonitor
+import org.eclipse.core.runtime.NullProgressMonitor
+import org.eclipse.emf.ecore.EObject
+import org.eclipse.ltk.core.refactoring.PerformChangeOperation
+import org.eclipse.swt.graphics.Image
+import org.eclipse.ui.actions.WorkspaceModifyOperation
+import org.eclipse.ui.views.markers.WorkbenchMarkerResolution
+import org.eclipse.xtend.lib.annotations.Accessors
+import org.eclipse.xtext.ide.serializer.impl.ChangeSerializer
+import org.eclipse.xtext.ui.editor.model.edit.IContextFreeModification
+import org.eclipse.xtext.ui.editor.model.edit.IContextFreeModification.PreInitializedModification
+import org.eclipse.xtext.ui.refactoring2.ChangeConverter
+import org.eclipse.xtext.ui.refactoring2.LtkIssueAcceptor
+import org.eclipse.xtext.ui.resource.IResourceSetProvider
+import org.eclipse.xtext.ui.util.IssueUtil
+
+/**
+ *  MarkerResolution which extends WorkbenchMarkerResolution and can be applied on multiple markers.
+ * 
+ * @author dhuebner - Initial contribution and API
+ * @since 2.13
+ */
+class WorkbenchMarkerResolutionAdapter extends WorkbenchMarkerResolution {
+
+	@Inject MarkerResolutionGenerator resolutionsGenerator
+	@Inject IssueUtil issueUtil
+	@Inject IResourceSetProvider resSetProvider
+	@Inject Provider<ChangeSerializer> serializerProvider
+	private final static Logger LOG = Logger.getLogger(WorkbenchMarkerResolutionAdapter);
+	@Accessors IssueResolution primaryResolution
+	@Accessors IMarker primaryMarker
+
+	override findOtherMarkers(IMarker[] markers) {
+		markers.filter[issueUtil.getCode(primaryMarker) == issueUtil.getCode(it)]
+	}
+
+	override String getLabel() {
+		return primaryResolution.getLabel()
+	}
+
+	override run(IMarker[] markers, IProgressMonitor monitor) {
+		new WorkspaceModifyOperation() {
+			override protected execute(
+				IProgressMonitor monitor) throws CoreException, InvocationTargetException, InterruptedException {
+				monitor.beginTask("Applying resolutions", markers.size)
+				markers.collectResolutions.forEach [
+					monitor.taskName = "Applying resolution"
+					run(it.key, it.value, monitor)
+					monitor.internalWorked(1)
+				]
+				monitor.done
+			}
+		}.run(monitor)
+	}
+
+	override void run(IMarker marker) {
+		if (!marker.exists) {
+			return
+		}
+		val resolutionData = marker.resolution
+		run(resolutionData.key, resolutionData.value, new NullProgressMonitor)
+	}
+
+	def collectResolutions(IMarker... markers) {
+		markers.map[marker|marker.resolution].filterNull.toList
```

Comment:
Can we make this cancellable in case of plenty of markers / expensive resolution computaiton?

Reply:
Similar to the CancelableResolutionAcceptor or using the progress monitor?

Reply:
I think using the progress monitor that is available in the super type on the run(..) method would be good enough - it should be checked after each marker.

Reply:
@szarnekow I was looking for a good test class to test the different kinds of quick-fix implementations, especially the Pre-Initialazed one, it should work with markers and custom QuickFixProvider implementation. Do you know a good one in xtext.ui.tests or xtext.xtext.ui.tests?

Reply:
Hey @dhuebner , unfortunately I'm not aware of a great utility to test quickfixes.